src/Microsoft.ML.Core/Utilities/ThreadUtils.cs

```diff
@@ -4,44 +4,154 @@
 
 using System;
 using System.Collections.Generic;
-using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 
 namespace Microsoft.ML.Internal.Utilities
 {
     internal static partial class Utils
     {
-        public static Thread CreateBackgroundThread(ParameterizedThreadStart start)
-        {
-            return new Thread(start)
-            {
-                IsBackground = true
-            };
-        }
+        public static Task StartBackgroundThread(Action start) =>
+            ImmediateBackgroundThreadPool.Queue(start);
 
-        public static Thread CreateBackgroundThread(ThreadStart start)
-        {
-            return new Thread(start)
-            {
-                IsBackground = true
-            };
-        }
+        public static Task StartBackgroundThread(Action<object> start, object obj) =>
+            ImmediateBackgroundThreadPool.Queue(start, obj);
 
-        public static Thread CreateForegroundThread(ParameterizedThreadStart start)
-        {
-            return new Thread(start)
-            {
-                IsBackground = false
-            };
-        }
+        public static Thread CreateForegroundThread(ParameterizedThreadStart start) =>
+            new Thread(start) { IsBackground = false };
 
-        public static Thread CreateForegroundThread(ThreadStart start)
+        /// <summary>
+        /// Naive thread pool focused on reducing the latency to execution of chunky work items as much as possible.
+        /// If a thread is ready to process a work item the moment a work item is queued, it's used, otherwise
+        /// a new thread is created.  This is meant as a stop-gap measure for workloads that would otherwise be
```

Comment:
>   [](start = 36, length = 2)

So, not a huge blocker, but if you read our comments you'll see that we generally follow the standard advice of one space, rather than two, after sentences. (E.g., [as described here](https://slate.com/technology/2011/01/two-spaces-after-a-period-why-you-should-never-ever-do-it.html).) I would perhaps put it a bit less insultingly than the author of that article, while of course agreeing with the general advice.

Reply:
Ok, will change.